src/autotype/AutoType.cpp

```diff
@@ -339,20 +339,22 @@ void AutoType::executeAutoTypeActions(const Entry* entry,
                 if (getMainWindow()) {
                     MessageBox::critical(getMainWindow(), tr("Auto-Type Error"), result.errorString());
                 }
-                emit autotypeRejected();
-                m_inAutoType.unlock();
-                return;
+                failed = true;
+                break;
             }
 
             Tools::wait(delay);
         }
-    }
 
-    m_windowForGlobal = 0;
-    m_windowTitleForGlobal.clear();
+        // Last action failed to complete, cancel the rest of the sequence
+        if (failed) {
+            break;
+        }
+    }
 
-    emit autotypePerformed();
+    resetAutoTypeState();
     m_inAutoType.unlock();
```

Comment:
This could be a QMutexLocker.

Reply:
QMutexLocker doesn't let you test for a lock, we test if it is already locked and bail out if it is instead of blocking.